### PR TITLE
[5.2] [Guided Tours] Addition of 'Auto start' step in 'How to create a tour?' tour

### PR DESCRIPTION
**Pull request by obuisard**
Pull Request for Issue #44094.

### Summary of Changes

The 'Auto Start' was added to the tour parameters.
The tour needs to take into account that parameter in adding a new step in the tour.

### Testing Instructions

You need to test on a fresh install and on update.
Run the tour 'How to create a tour?'.

### Actual result BEFORE applying this Pull Request

The tour runs and bypasses the parameter 'Auto Start'.

### Expected result AFTER applying this Pull Request

The tour runs and has a step for the 'Auto Start' parameter, just before the step 'Save and close'.

autostart

### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

**Comment by obuisard** [comment]
Richard @richard67 I am not sure about the updated scripts.

I cannot add a step in a tour between other steps and ensure the steps are in the right order.
Therefore I deleted all steps for the tour, then re-introduced them. 
I don't see another way to do this.

Also, I am not sure it's good practice to use UNION ALL in the INSERT INTO ... SELECT.

**Comment by obuisard** [comment]
Brian @brianteeman How is the wording for you?

**Comment by richard67** [comment]
> Richard @richard67 I am not sure about the updated scripts.

@obuisard I can’t check before weekend.

**Comment by richard67** [comment]
> Richard @richard67 I am not sure about the updated scripts.
> 
> I cannot add a step in a tour between other steps and ensure the steps are in the right order. Therefore I deleted all steps for the tour, then re-introduced them. I don't see another way to do this.

@obuisard Does that mean the tour steps need to be saved in database in the same order as they appear in the list because they obviously don't have an `order` column? As far as I can see that's the case, and if so, then indeed there is no way to insert a record somewhere in the middle. But to be honest: That's poor design. It always should be possible to add a step in the middle, and the right way to do that would be to use an `ordering` column so that ou always can reorder steps or add steps in the middle.

**Comment by richard67** [comment]
> The order of steps is recorded in the `ordering` column.
> 
> In the event the order of the steps had not been changed since install, it would go between steps with ids 5 and 6 (which have an ordering of 5 and 6 respectively). Would give the new step an order of 5 be enough?

@obuisard I did not see the ordering column being used in your insert statement. That’s why I was not aware of it.

when inserting something new between ordering 5 and 6, you can first update all steps with ordering >= 6 to set ordering = ordering + 1, so what had 6 before will have 7, what had 7 before will have 8 and so on, and ordering 6 would be free. Then insert the new step with ordering 6.

In general, when you do an „update sometable set somevalue=somevalue+1 where somecondition“, you increment somevalue by 1 where somecondition is true.

**Comment by obuisard** [comment]
> > The order of steps is recorded in the `ordering` column.
> > In the event the order of the steps had not been changed since install, it would go between steps with ids 5 and 6 (which have an ordering of 5 and 6 respectively). Would give the new step an order of 5 be enough?
> 
> @obuisard I did not see the ordering column being used in your insert statements in the update SQL scripts. That’s why I was not aware of it.
> 
> when inserting something new between ordering 5 and 6, you can first update all steps with ordering >= 6 to set ordering = ordering + 1, so what had 6 before will have 7, what had 7 before will have 8 and so on, and ordering 6 would be free. Then insert the new step with ordering 6.
> 
> In general, when you do an „update sometable set somevalue=somevalue+1 where somecondition“, you increment somevalue by 1 where somecondition is true.

Thanks, Richard

File: administrator/components/com_admin/sql/updates/mysql/5.2.0-2024-09-17.sql [hunk 1]
```diff
@@ -0,0 +1,12 @@
+UPDATE `#__guidedtour_steps`
+SET `ordering` = `ordering` + 1
+WHERE `title` = 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_SAVECLOSE_TITLE';
+
+UPDATE `#__guidedtour_steps`
+SET `ordering` = `ordering` + 1
+WHERE `title` = 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_CONGRATULATIONS_TITLE';
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/mysql/5.2.0-2024-09-17.sql]:
```suggestion
WHERE `title` IN ('COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_SAVECLOSE_TITLE', 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_CONGRATULATIONS_TITLE');
```
Why 2 statements where 1 is enough?

File: administrator/components/com_admin/sql/updates/postgresql/5.2.0-2024-09-17.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+UPDATE "#__guidedtour_steps"
+SET "ordering" = "ordering" + 1
+WHERE "title" = 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_SAVECLOSE_TITLE';
+
+UPDATE "#__guidedtour_steps"
+SET "ordering" = "ordering" + 1
+WHERE "title" = 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_CONGRATULATIONS_TITLE';
```
Comment (richard67) [inline administrator/components/com_admin/sql/updates/postgresql/5.2.0-2024-09-17.sql]:
```suggestion
WHERE "title" IN ('COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_SAVECLOSE_TITLE', 'COM_GUIDEDTOURS_TOUR_GUIDEDTOURS_STEP_CONGRATULATIONS_TITLE');
```
Why 2 statements where 1 is enough?

File: administrator/components/com_admin/sql/updates/mysql/5.2.0-2024-09-17.sql (hunk 1 above)
Comment (richard67) [reply administrator/components/com_admin/sql/updates/mysql/5.2.0-2024-09-17.sql]: P.S.: The ordering = ordering + 1 works for each record individually if you update multiple records at once.

**Comment by brianteeman** [comment]
technically this update query would not be correct if the user has changed the order of the steps but as thats very unlikely I can confirm the update sql works on mysql

**Comment by Kostelano** [comment]
No, something is not working.
As soon as you complete a tour (create a new one) and enable automatic start, the site will stop working. You can enable ALL components (then the start will stop working immediately) or you can select a specific component - then when you go to the component page, the site will stop working (404).

All newly created tours are placed in the Dashboard, although I have installed other components.

Screenshot_1

**Comment by obuisard** [comment]
> No, something is not working. As soon as you complete a tour (create a new one) and enable automatic start, the site will stop working. You can enable ALL components (then the start will stop working immediately) or you can select a specific component - then when you go to the component page, the site will stop working (404).
> 
> All newly created tours are placed in the Dashboard, although I have installed other components.

I have installed the PR as a fresh install, and as an update, created new tours, with auto starting enabled and did not seem to  encounter any of those issues.
Was your install a fresh install? An update? Did you see any Database warning in the console?

The 404 error you are getting: what is the URL in your browser for that page? Could it be that the URL parameter you have set in the tour is incorrect?

The tours you have created: are 123, 234, 455 the titles of the tours?

If you left 'Component selector' set to 'all', the tours will show in the Dashboard section (in the list of all the tours). Although maybe, as an improvement, we should put those special cases under no section... But it is unrelated to this PR.

image

If you create a tour for banners, you would select 'Banners' for the tour to show under the 'Banners' section.

If you have 3 tours set to auto-start and set to start anywhere (= when the component selector is set to 'all'), any refresh of any console page will start a new tour after you have either completed, cancelled or skipped the previous one.

**Comment by Kostelano** [comment]
I have tested this item :white_check_mark: successfully on 1bb95a103937bcd8c58e19a30485d61af2c49ce0<br /><br />I found a bug (an extra character in the URL that threw me off my feet), so I apologize for making you waste time looking for a non-existent bug, it happens ;).

Well, I tested PR with the old installation + service pack with PR and separately with the new installation. No errors were noticed, below I am posting a successful test. Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44105">issues.joomla.org/tracker/joomla-cms/44105</a>.</sub>

**Comment by exlemor** [comment]
I have tested this item :white_check_mark: successfully on 1bb95a103937bcd8c58e19a30485d61af2c49ce0<br /><br />I have tested this item ✅ successfully.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44105">issues.joomla.org/tracker/joomla-cms/44105</a>.</sub>

**Comment by richard67** [comment]
RTC <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44105">issues.joomla.org/tracker/joomla-cms/44105</a>.</sub>

**Comment by richard67** [comment]
@alikon The PR has 2 good tests so I've set it RTC, but it would be good to get an additional quick test with PostgreSQL, too, just to see if installation with and updating to the patched package of this PR works. I would be happy if you could find the time. Thanks in advance.

**Comment by alikon** [comment]
update from 5.1.4 to  Prebuilt package of this pr doesn't work

image

`INSERT INTO "#__action_logs_extensions" ("extension") VALUES ("com_guidedtours");`  coming from #43814

**Comment by richard67** [comment]
@obuisard On PostgreSQL, double quotes are for names (table names, column names) but not for strings. Strings are quoted with single quotes.

**Review by alikon — APPROVED** [review]
this pr as "per se" works fine on postgres

**Comment by obuisard** [comment]
> @obuisard On PostgreSQL, double quotes are for names (table names, column names) but not for strings. Strings are quoted with single quotes.

Thanks Richard @richard67 , yes, I missed it. Thanks Nicola @alikon for the correction.

**Comment by Hackwar** [comment]
Thank you for your contribution @obuisard!

**Pull request merged**